DoFTools::make_flux_sparsity_pattern with masks, in parallel

spco opened the pull request:
Here's my work on a parallelised version of make_flux_sparsity_pattern() with masks on the internal and face dofs. (I have the necessary changes working in ASPECT as discussed with @bangerth etc for this to be used there too, as a solution to https://github.com/geodynamics/aspect/issues/1112, but I'll hold off on that until this has merged.)

I'll squash it later once review is complete.

drwells commented:
This looks good. My main concern is that most of the code in this function is shared with the non-hp `make_flux_sparsity_pattern` function already in the file. Would it be possible, since that is already an internal function, to add more optional arguments to it instead of writing a new function?

spco commented:
That is almost certainly possible - I'll investigate and get back!

spco commented:
Ok, how does this version look? The previously available function 

```
DoFTools::make_flux_sparsity_pattern (dof_handler,
                                      sp,
                                      coupling,
                                      face_coupling);
```

now forwards to 

```
DoFTools::make_flux_sparsity_pattern (dof_handler,
                                      sp,
                                      constraints, keep_constrained_dofs,
                                      coupling,
                                      face_coupling,
                                      subdomain_id);
```
with default `ConstraintMatrix` and `keep_constrained_dofs`, and `subdomain_id` has a default.

The one assumption I have made is that
```
for (unsigned int i=0; i<fe.dofs_per_cell; ++i)
  for (unsigned int j=0; j<fe.dofs_per_cell; ++j)
    if (int_dof_mask (i,j) != none)
      sparsity.add (dofs_on_this_cell[i],
                    dofs_on_this_cell[j]);
```
is equivalent to 
```
constraints.add_entries_local_to_global (dofs_on_this_cell,
                                         sparsity,
                                         keep_constrained_dofs,
                                         int_dof_mask[cell->active_fe_index()]);
``` 
with a default `ConstraintMatrix constraints;` and `keep_constrained_dofs=true`

This seems like a sensible conclusion, but I'd love a second opinion - the codebase for constraints and sparsity patterns is like a warren for anyone just dipping in and out!

spco commented:
It might also make sense to split off the functionality inside the `if (need_dof_mask == true)` section, as this is used in 3 places in the file now, rather than just once. Thoughts?

drwells commented:
It looks fine at a glance, but the sparsity pattern code is surprisingly tricky (if you look at the history of this file you will see that I have both broken and fixed this very function) so I will reserve judgement until I see some new tests :)

Yes, those two approaches should be the same according to the documentation.

Yes, the hp version is duplicated. It would be nice to refactor that out.

I can't seem to compile your changes: it looks like the library tries to create instantiations for the old internal function still. Could you take a look at that?

spco commented:
The extra instantiation should now be fixed - would you mind retrying when you have a moment?

spco commented:
I've pulled that repeated code out into a new function. This should be good for review, if it compiles for you this time!

In terms of the tests, I think the affected functions are called in 
tests/integrators/distribute_flux_sparsity_pattern.cc
tests/lac/constraints_block_01.cc
tests/bits/dof_tools_18a.cc   (moved now to tests/dofs)
tests/bits/dof_tools_18b.cc   (moved now to tests/dofs)
tests/bits/dof_tools_18c.cc   (moved now to tests/dofs)
tests/bits/dof_tools_18d.cc   (moved now to tests/dofs)
tests/trilinos/elide_zeros.cc

This should check that we don't break the existing functionality. The unknown quantities are:
1) m_f_s_p with a combination of a non-empty ConstraintMatrix and non-empty int/flux_mask
2) the same on multiple processes
3) m_f_s_p with an empty Constaint/Matrix and a non-empty int/flux_mask
4) the same on multiple processes

Does that sound reasonable? I've never contributed tests to dealii directly (I have to ASPECT) so is there a place to look for an explanation?

drwells commented:
@spco I am traveling for the next few days so I won't be able to look at this in detail, but I think I can answer your questions.

I don't know if there is a guide for adding tests, but its not that hard to do: you just need to add a test to the right directory and a `output` file with the expected output. The files you listed are good examples.

bangerth commented:
@spco -- information on testing is at `doc/users/testsuite.html` and `doc/developers/testsuite.html` in the deal.II directories. It's actually all rather simple if you've done it a couple of times.

bangerth commented:
As for tests: there is `tests/mpi/periodicity_02.cc` that calls the non-DG version `make_sparsity_pattern` in parallel, as does the _03 test. You could take that as a starting point for tests of your function (though it might make it too complicated to have periodicity for your DG function, unless it already handles this correctly). In general, I almost always clone existing tests when I want to write new ones, and then add or subtract what it is I specifically want to test.

spco commented:
Thanks - that documentation looks pretty comprehensive! I'll take a look tomorrow.

bangerth commented on include/deal.II/dofs/dof_tools.h:
```diff
@@ -645,10 +645,25 @@ namespace DoFTools
+  /**
+   * This function does essentially the same as the previous
+   * make_flux_sparsity_pattern() function but allows the application of
+   * a constraint matrix.
```
Maybe add that this is useful in cases where *some* of the components of a finite element are continuous whereas others are discontinuous (and consequently need fluxes in the sparsity pattern). I think this would omit the otherwise understandable question of "why would I want to do this for DG elements?".

bangerth commented on include/deal.II/dofs/dof_tools.h:
```diff
@@ -2260,6 +2275,33 @@ namespace DoFTools
+  template <int dim, int spacedim>
+  Table<2,bool>
+  bool_dof_couplings_from_component_couplings (const FiniteElement<dim,spacedim> &fe,
+                                               const Table<2,Coupling> &mask_in);
```
I don't think I quite understand the description or purpose of this function. Is this function essentially "flattening" the `Table<2,Couplings>` to `Table<2,bool>` by converting each entry to a boolean by comparing `!= none`?

bangerth commented on source/dofs/dof_tools_sparsity.cc:
```diff
@@ -708,6 +667,84 @@ namespace DoFTools
+    for (unsigned int i=0; i<mask_in.n_rows(); ++i)
+      for (unsigned int j=0; j<mask_in.n_cols(); ++j)
+        if (mask_in(i,j) == none)
+          need_dof_mask = true;
```
...at which point you could just jump out of the double loop :-)

bangerth replied:
but why do you compare against `none`? can you elaborate in a comment when exactly you want to compute the DoF mask, given the different possibilities you might encounter in `mask_in`'s entries?

bangerth commented on source/dofs/dof_tools_sparsity.cc:
```diff
@@ -708,6 +667,84 @@ namespace DoFTools
+
+    const unsigned int n_dofs = fe.dofs_per_cell;
+
+    Table<2,bool> mask_out(n_dofs, n_dofs);
```
how are these values initialized?

bangerth commented on source/dofs/dof_tools_sparsity.cc:
```diff
@@ -708,6 +667,84 @@ namespace DoFTools
+      {
+        const unsigned int dofs_per_cell = fe.dofs_per_cell;
+
+        mask_out.reinit (dofs_per_cell, dofs_per_cell);
```
you've already set the size above. no need to do it twice

bangerth commented on source/dofs/dof_tools_sparsity.cc:
```diff
@@ -708,6 +667,84 @@ namespace DoFTools
+
+    if (need_dof_mask == true)
+      {
+        const unsigned int dofs_per_cell = fe.dofs_per_cell;
```
this equals `n_dofs` above.

bangerth commented on source/dofs/dof_tools_sparsity.cc:
```diff
@@ -708,6 +667,84 @@ namespace DoFTools
+                fe.is_primitive(j))
+              mask_out(i,j)
+                = (mask_in(fe.system_to_component_index(i).first,
+                           fe.system_to_component_index(j).first) != none);
```
This and line 723 is what I don't understand in my comments above. The rest is otherwise actually quite clear.

bangerth commented on source/dofs/dof_tools_sparsity.cc:
```diff
@@ -738,19 +777,23 @@ namespace DoFTools
               // make sparsity pattern for this cell
-              for (unsigned int i=0; i<fe.dofs_per_cell; ++i)
-                for (unsigned int j=0; j<fe.dofs_per_cell; ++j)
-                  if (int_dof_mask (i,j) != none)
```
ah, I see that this is where the comparison is from. Why not build the regular mask and just convert it by hand here?

spco commented:
I've improved the comments, and I've removed `bool_dof_couplings_from_component_couplings` as this was just reusing code from `dof_couplings_from_component_couplings` - it's more sane just to use the latter and then convert by hand to `bool`. Hopefully this addresses all the comments.

bangerth commented on source/dofs/dof_tools_sparsity.cc:
```diff
@@ -145,51 +145,22 @@ namespace DoFTools
+    std::vector<Table<2,Coupling> > dof_mask (fe_collection.size());
+
+    dof_mask = dof_couplings_from_component_couplings (fe_collection,
+                                                       couplings);
```
I'd say combine these two statements and make `dof_mask` a `const` variable.

bangerth commented:
Looks good to merge if (i) we can get a test or two, (ii) you can merge the two lines I mentioned in a comment, (iii) you squash the 15 commits into one. Let me know if you need help with any of these steps!

Other than that, this is great. Thanks for attacking this issue!

bangerth commented:
Oh, (iv) make sure you add an entry to `doc/news/changes.h` in the "specific improvements" section, so you get the credit you deserve. There have probably been other patches that were merged since you started and that added entries to this file, so I guess you'll want to rebase to master.

bangerth commented:
I think only tests are missing. Let me know if you need help!

spco commented:
Yes, I have the testsuite up and running here, but am not sure exactly how best to test this functionality. Is it enough to take an `FESystem` of a conforming element and nonconforming element, form a small mesh (square, only a few cells but ensuring we have hanging nodes), form the sparsity pattern on this, and then print/partially print?

Lots of the tests seem to involve much more of the assembly/solve cycle than this - should that also be included?
And would it be better to have multiple tests covering different combinations? For example, `dofs/dof_tools_18*.cc` cycle over different sparsity matrix types.

spco commented:
Ok, I've added a test, but it doesn't work. Perhaps you could give it a try? The output in the supplied .output files is garbage at this point, but that doesn't matter. The point is that when I run this test on multiple MPI processes, the debug and release outputs from `sp.print_gnuplot()` don't match each other. Mostly they're fine, but there are sections of output that are clearly junk - zeros, or large numbers. 

Note that this happens both when I use the new `m_f_s_p` function and when, as in the latest commit, I comment out the masks so I'm calling an untouched `m_f_s_p` function, which should't be touched by this PR.

I'm going a little crazy trying to spot what I've failed to initialize. Either that, or the behaviour of sparsity patterns doesn't match my expectations. Perhaps both! Any suggestions?

bangerth commented on tests/dofs/dof_tools_22a.cc:
```diff
@@ -0,0 +1,160 @@
+//
+// ---------------------------------------------------------------------
+//
+// This tests was written by Sam Cox.
```
Can you add here what the test does?

bangerth commented on tests/dofs/dof_tools_22a.cc:
```diff
@@ -0,0 +1,160 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2001 - 2016 by the deal.II authors
```
wrong year

bangerth commented on tests/dofs/dof_tools_22a.cc:
```diff
@@ -0,0 +1,160 @@
+    
+    std::ofstream logfile("output");
+    deallog.attach(logfile);
+    deallog.threshold_double(1.e-10);
```
Here, *all* processes are opening the `output` file, and probably reset it to zero length. What happens if you remove these three lines from the function?

bangerth replied:
In particular, I don't think you need it because you only write into this file from processor 0, and that processor has already opened it in `main()`.

bangerth commented on tests/dofs/dof_tools_22a.cc:
```diff
@@ -0,0 +1,160 @@
+    file << "My locally owned dofs: ";
+    dh.locally_owned_dofs().print(file);
+    file << std::endl;
+    MPI_Barrier(MPI_COMM_WORLD);
```
need to `close` the file before the barrier.

bangerth commented on tests/dofs/dof_tools_22a.cc:
```diff
@@ -0,0 +1,160 @@
+}
+
+
+//
```
endline missing

spco commented:
Thanks for the comments. They should all be addressed.

So this test passes for a single MPI process but fails for multiple. For example, from the output for 3 processes:
lines 6675-6719 in debug mode:
```
51 0
52 0
53 0
54 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
655360 0
92 -1
93 -1
94 -1
95 -1
```
and in release mode:
```
51 0
52 0
53 0
54 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
0 0
92 -1
93 -1
94 -1
95 -1
```

There is a similar discrepancy in lines 5590-5638, characterised by zeros in release mode, and a large number in debug.

On 5 processes (which is too many really - proc 4 doesn't have any dofs) the only discrepancy is lines 6550-6580: Debug:
```
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
0 -14
78 -14
78 -14
78 -14
78 -14
78 -14
78 -14
```
Release:
```
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
78 -13
1 -14
2 -14
3 -14
4 -14
5 -14
6 -14
7 -14
8 -14
9 -14
10 -14
11 -14
12 -14
13 -14
14 -14
15 -14
78 -14
78 -14
78 -14
78 -14
78 -14
78 -14
```

I'm not sure which is the correct question to be asking: why do we see these discrepancies? Or perhaps more pertinently, why is `write_gnuplot()` outputting repeated values, like `78 -14`? For example, on 3 processors, the last processor finishes with
```
56 -53
61 -53
75 -53
76 -53
104 -53
106 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -53
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54
0 -54

My locally owned dofs: {[92,146]}
```
I feel like the second is probably closer to the root cause.

tjhei commented on tests/dofs/dof_tools_22a.cc:
```diff
@@ -0,0 +1,157 @@
+  deallog.push(Utilities::int_to_string(myid));
+
+  std::ofstream file((std::string("dat.") + Utilities::int_to_string(myid)).c_str());
+  file << "**** proc " << myid << ": \n\n";
```
This is way to complicated to generate parallel output. You can just use ``MPILogInitAll log;`` in ``main`` and output from every processor will be copied into the output correctly.

tjhei commented:
Have you tried running under ``valgrind`` to see if you are overwriting memory somewhere?

spco commented:
Hi @tjhei. I ran the test under `valgrind` and the first error was 
```
==32097==  Address 0x118b49340 is 0 bytes after a block of size 19,088 alloc'd
==32097==    at 0x1000A0EBB: malloc (in /usr/local/Cellar/valgrind/3.11.0/lib/valgrind/vgpreload_memcheck-amd64-darwin.so)
==32097==    by 0x1173247DD: operator new(unsigned long) (in /usr/lib/libc++.1.dylib)
==32097==    by 0x11625ADE3: Epetra_IntSerialDenseMatrix::Shape(int, int) (in /Applications/deal.II.app/Contents/Resources/opt/trilinos-0b08cd5/lib/libepetra.12.6.1.dylib)
==32097==    by 0x1161F1590: Epetra_CrsGraph::OptimizeStorage() (in /Applications/deal.II.app/Contents/Resources/opt/trilinos-0b08cd5/lib/libepetra.12.6.1.dylib)
==32097==    by 0x10A1576BB: dealii::TrilinosWrappers::SparsityPattern::compress() (trilinos_sparsity_pattern.cc:810)
==32097==    by 0x10001F873: test() (dof_tools_22a.cc:111)
==32097==    by 0x10001FC11: main (dof_tools_22a.cc:153)
```

I've changed to use a `DynamicSparsityPattern` and `SparsityTools::distribute_sparsity_pattern()` and everything now passes. 

I've left the relevant code in there which broke, in case anyone can tell me what I was doing wrong - I was probably `compress()`ing where I shouldn't or shouldn't.

Other than that, I'm happy to remove those lines and then the test is done.

spco commented:
For the sake of completeness, the other errors from `valgrind` were multiple variations on 
```
==32424== Conditional jump or move depends on uninitialised value(s)
==32424==    at 0x11756D8A7: __vfprintf (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x1175936C0: __v2printf (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x117577B5D: vsprintf_l (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x117567F83: sprintf_l (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x117312D39: std::__1::num_put<char, std::__1::ostreambuf_iterator<char, std::__1::char_traits<char> > >::do_put(std::__1::ostreambuf_iterator<char, std::__1::char_traits<char> >, std::__1::ios_base&, char, long) const (in /usr/lib/libc++.1.dylib)
==32424==    by 0x1172F9060: std::__1::basic_ostream<char, std::__1::char_traits<char> >::operator<<(int) (in /usr/lib/libc++.1.dylib)
==32424==    by 0x10A15DA41: dealii::TrilinosWrappers::SparsityPattern::print_gnuplot(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) const (trilinos_sparsity_pattern.cc:1126)
==32424==    by 0x10001F96F: test() (dof_tools_22a.cc:121)
==32424==    by 0x10001FC11: main (dof_tools_22a.cc:153)
``` 
(all within the `<<` call at `trilinos_sparsity_pattern.cc:1126`) and
```
==32424== Syscall param write(buf) points to uninitialised byte(s)
==32424==    at 0x117674612: write$NOCANCEL (in /usr/lib/system/libsystem_kernel.dylib)
==32424==    by 0x1175681F9: _swrite (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x117560724: __sflush (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x117560680: fflush (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x100027850: std::__1::basic_filebuf<char, std::__1::char_traits<char> >::sync() (fstream:870)
==32424==    by 0x1172F3F59: std::__1::basic_ostream<char, std::__1::char_traits<char> >::flush() (in /usr/lib/libc++.1.dylib)
==32424==    by 0x1048C86F7: std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::endl<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) (ostream:1035)
==32424==    by 0x10A15DA78: dealii::TrilinosWrappers::SparsityPattern::print_gnuplot(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) const (ostream:322)
==32424==    by 0x10001F96F: test() (dof_tools_22a.cc:121)
==32424==    by 0x10001FC11: main (dof_tools_22a.cc:153)
==32424==  Address 0x1190ed510 is 0 bytes inside a block of size 4,096 alloc'd
==32424==    at 0x1000A0EBB: malloc (in /usr/local/Cellar/valgrind/3.11.0/lib/valgrind/vgpreload_memcheck-amd64-darwin.so)
==32424==    by 0x117563AEE: __smakebuf (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x117578B41: __swsetup (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x117562D4A: __sfvwrite (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x117563342: fwrite (in /usr/lib/system/libsystem_c.dylib)
==32424==    by 0x100028917: std::__1::basic_filebuf<char, std::__1::char_traits<char> >::overflow(int) (fstream:689)
==32424==    by 0x1172F31CC: std::__1::basic_streambuf<char, std::__1::char_traits<char> >::xsputn(char const*, long) (in /usr/lib/libc++.1.dylib)
==32424==    by 0x1048D85AC: std::__1::ostreambuf_iterator<char, std::__1::char_traits<char> > std::__1::__pad_and_output<char, std::__1::char_traits<char> >(std::__1::ostreambuf_iterator<char, std::__1::char_traits<char> >, char const*, char const*, char const*, std::__1::ios_base&, char) (streambuf:360)
==32424==    by 0x1048D8304: std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) (ostream:755)
==32424==    by 0x1048D6B82: std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::operator<< <char, std::__1::char_traits<char>, std::__1::allocator<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) (in /Users/sam/git/dealii/build/lib/libdeal_II.g.8.5.0-pre.dylib)
==32424==    by 0x10A40CAB6: dealii::LogStream::attach(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, bool) (logstream.cc:227)
==32424==    by 0x10003768D: MPILogInitAll::MPILogInitAll(bool) (tests.h:356)
```

tjhei commented:
> I've changed to use a DynamicSparsityPattern and SparsityTools::distribute_sparsity_pattern() and everything now passes.
> 
> I've left the relevant code in there which broke, in case anyone can tell me what I was doing wrong - I was probably compress()ing where I shouldn't or shouldn't.

Huh, this looks like a bug. Can you open a separate issue for this please? Maybe @kronbichler has an idea what is going on.

tjhei commented:
> Other than that, I'm happy to remove those lines and then the test is done.

Yes, please do that so we can merge this PR.

spco commented:
Ok - edited, squashed, and good to go!

kronbichler commented:
> Huh, this looks like a bug. Can you open a separate issue for this please? Maybe @kronbichler has an idea what is going on.

@spco @tjhei I didn't follow this PR closely, but indeed it looks like there could be something wrong and a small test case would be appreciated.

bangerth commented:
Great, thanks!
